backend/app/app/crud/crud_studyset.py

```diff
@@ -247,8 +247,24 @@ def create_new_study_set(
 
         print('Final repetition model:', repetition_model)
         
-        if repetition_model == schemas.Repetition.karl and not is_mnemonic_deck:
-            eligible_facts = crud.fact.get_eligible_facts(query=base_facts_query, limit=send_limit, randomize=True)
+        eligible_facts = []
+        if repetition_model == schemas.Repetition.karl:
+            if is_mnemonic_deck:
+                new_facts_query = crud.helper.filter_only_new_facts(query=base_facts_query, user_id=user.id, log_type=schemas.Log.study)
+                new_eligible_facts = crud.fact.get_eligible_facts(query=new_facts_query, limit=send_limit, randomize=True)
+                old_eligible_facts = []
+                if len(new_eligible_facts) < return_limit:
+                    old_facts_query = crud.helper.filter_only_reviewed_facts(query=base_facts_query, user_id=user.id, log_type=schemas.Log.study)
+                    old_eligible_facts = crud.fact.get_eligible_facts(query=old_facts_query, limit=send_limit, randomize=True)
+
+                if len(new_eligible_facts) == 0:
+                    eligible_facts = old_eligible_facts
+                # only in this study session will we force users to study certain old facts
```

Comment:
yes! just for the special case when a user requests 20 facts and there are only 15 new facts, for example. we would need 5 old facts to get a set of 20, but rather than re-running the entire scheduler from Shi's side, I just select 5 old facts since this will only happen once